### ISSUE TITLE
Pre-Remove callback issues

### ISSUE THREAD

**Issue by robludwig** [Automattic/mongoose#3026]
I've tested this with mongoose v. 3.8.27 and v  4.0.3

I'm encountering some strange issues with pre-remove middleware. My goal is to add pre-remove middlware that prevents some documents from being removed if certain conditions are true. My understanding is that this can be done by calling the `next` callback in serial middleware with an error parameter, which should short-circuit the middleware chain, and prevent the document from being removed.

What I've found is a little strange: if I use the function signature `Schema.pre('remove', function(next)` to write my middleware function, errors are *not* propogated and the document is removed anyway. Additionally, in 3.8.27 calling the `next()` function doesn't terminate the middleware chain, and leads to blocking.

Am I misunderstanding the goal of pre-remove middleware? The [documentation](http://mongoosejs.com/docs/middleware.html) says "If any middleware calls next or done with an error, the flow is interrupted, and the error is passed to the callback." Am I mistaken in thinking that should prevent my document from being removed?

Thanks for any help/guidance,
-Rob

**Comment by robludwig** [comment]
Below is a sample which demonstrates the issue:
```
/**
 * Created by rob on 5/28/15.
 */
"use strict";
var mongoose = require('mongoose');
var Schema = mongoose.Schema;
var chai = require('chai');
var expect = chai.expect;
var async = require('async');

var NumbertestDef = {
    num : {type : Number, required : true}
}
var NumbertestSchema = new Schema(NumbertestDef);


/*
NumbertestSchema.pre('remove', function(next){
    var doc = this;
    console.log("removing doc with num", doc.num);
    if(doc.num > 10) return next();
    return next({error : "NumTooSmall", message : "That number is too small to be removed"});
});
*/

NumbertestSchema.pre('remove', false, function(next, done){
    var doc = this;
    console.log("removing doc with num", doc.num);
    if(doc.num > 10) return next();
    next({error : "NumTooSmall", message : "That number is too small to be removed"});
    return done({error : "NumTooSmall", message : "That number is too small to be removed"})
});

var Numbertest = mongoose.model("Numbertest", NumbertestSchema);

describe("Numbertest", function(){
    var nums;
    before(function(done){
        Numbertest.remove(done)
    })
    before(function(done){
        nums = [1, 2, 10, 3, 11, 30];
        async.each(nums, function(num, cb){
            new Numbertest({num : num}).save(cb)
        }, done);
    })
    it("should call the callback in the normal deletion case", function(done){
        Numbertest.findOne({num : 11}, function(err, nt){
            nt.remove(done);
        })
    })
    it("should call the callback with an error in the error case", function(done){
        Numbertest.findOne({num : 1}, function(err, nt){
            nt.remove(function(err){
                expect(err).to.not.be.null;
                Numbertest.count(function(err, count){
                    expect(count).to.equal(nums.length - 1); //we've deleted one..
                    done()
                })
            })
        })
    })
})




```

**Comment by robludwig** [comment]
This may be a duplicate of #2441

**Comment by robludwig** [comment]
It is, changing the above code to callback with an `Error()` instead of an object causes the expected behavior. In the discussion for #2441 it was mentioned that "This is fixed in 3.9.x". Is that not the case, or was the fix only for strings and not objects?

**Comment by robludwig** [comment]
Alternatively, the documentation cited above which states "If any middleware calls next or done with an error, the flow is interrupted, and the error is passed to the callback." could be edited to make clear that one need to call the callback with an `Error` and not simply any error parameter.

**Comment by vkarpov15** [comment]
In theory this was fixed for 4.0.0, but you're right, in 3.8.x you **need** to pass something that's of type `Error`. I'll investigate further later as to why this is still happening for you in 4.0.x, because it definitely should not.

**Comment by vkarpov15** [comment]
Sorry for postponing this issue for so long. Unfortunately this was not properly fixed in 3.9, because we still support behavior like [this](https://github.com/Automattic/mongoose/blob/98445926a994f9fff8372a47d9702bc4fd2477b2/test/document.hooks.test.js#L220-L230). See #3672.
